### ISSUE TITLE
Retry command fails and gets stuck inside a loop

### ISSUE THREAD

**Issue by cloudowski** [openshift/jenkins-client-plugin#209]
We have the following piece of code in our pipeline:

```
timeout(time: 3, unit: 'MINUTES') {
                    podSelector.untilEach {
                        echo "pod: ${it.name()}"
                        return it.object().status.containerStatuses[0].ready
                    }
                }
```

It checks whether all containers are ready. In most cases, it's a single container, but sometimes our job fails and it seems like it's not retrying check of the container status - it seems like it's not entering loop.
We get the following output:

```
11:39:36 Checking watch output again in 300ms
11:39:36 Checking watch output again in 360ms
11:39:36 Checking watch output again in 432ms
11:39:37 Checking watch output again in 518ms
11:39:37 Checking watch output again in 621ms
11:39:38 Checking watch output again in 745ms
11:39:39 Checking watch output again in 894ms
11:39:40 Checking watch output again in 1072ms
11:39:41 Checking watch output again in 1286ms
11:39:42 Checking watch output again in 1543ms
11:39:43 Checking watch output again in 1851ms
11:39:45 Checking watch output again in 2221ms
11:39:47 Checking watch output again in 2665ms
11:39:50 Checking watch output again in 3198ms
11:42:18 Cancelling nested steps due to timeout
```

It should retry in 3.2s, but it didn't and it finally timed out. It sometimes work, but recently we noticed increased number of such errors.
We use version 1.0.23.

**Comment by szpak** [comment]
We took a closer look at the issue and the problem appears on reading data from `/tmp/watchstderr...txt` file. When to the initial content of that file (e.g. `pod/foobar-28-nxsjq`) the same  line is added on deployment completion the execution seems to hang on:
```
byte[] newOutput = Files.readAllBytes(path);
```
It's terminated by our timeout a few minutes later. The deployment itself finishes successfully.

Looking at the recent changes it seems to be a regression introduced in https://github.com/openshift/jenkins-client-plugin/commit/11ba99797511157f9c8dfea18bfec6f34ed6edfc ([bug 1657208](https://bugzilla.redhat.com/show_bug.cgi?id=1657208)) (1.0.23). The problem was not observable with 1.0.22. 

@gabemontero Maybe you have any thought on that?

Jenkins 2.121.3. OpenShift 3.11.16.

**Comment by gabemontero** [comment]
Thanks for the additional detail @szpak 

Yeah we had to switch the lib used to read files with the fixes introduced in 1.0.23.  We have not seen the hang you noted in our watch related automated testing, but I certainly could see such a thing being intermittent or dependent on the precise output produced by the underlying `oc` command.

I'll have to investigate the underlying implementations of the file libs used before and after the 1.0.23 change and see what workarounds are possible.

Certainly if you don't need to the fix provided in 1.0.23 (which revolves around use of the k8s plugin container step for pipelines, where multi-container pods are used), use 1.0.22 in the interim.

**Comment by szpak** [comment]
@gabemontero Thanks for your reply. I downgraded to 1.0.22 and it seems to work fine, but I had to workaround issues which I didn't observe in 1.0.23. Such as:

```
def dcObj = openshift.selector('dc', env.APP_NAME).object()
def podSelector = openshift.selector('pod', [deployment: "${APP_NAME}-${dcObj.status.latestVersion}"])
podSelector.untilEach {
    return it.object().status.containerStatuses[0].ready
}
assert podSelector.object().status.containerStatuses[0].ready //fails
```
Nevertheless, it's not related to that issue and `untilEach` can be put in a loop and the second iteration passes then. I will check it in 1.0.24 once available and eventually report a separate ticket.

Please let me know if you need anything else from my side regarding the hanging on read.

**Comment by severi** [comment]
experiencing a similar issue with this:
```
def replicationControllerSelector = openshift.selector("rc/my-api-123")

replicationControllerSelector.watch {
    rplc = it.object()
    if (rplc.metadata.annotations.'openshift.io/deployment.phase' == 'Failed') {
        error("error")
        return true
    }
    if (rplc.metadata.annotations.'openshift.io/deployment.phase' == 'Complete') {
        return true
    }
    return false
}
```
When checking the rc-resource manually  I can see that the phase is "Complete" but for some reason this doesn't trigger the the closure (or actually it does after waiting for ~2,5hours). No issues regarding this with the previous versions

**Comment by gabemontero** [comment]
fyi although it has proven very intermittent for us, I believe we have seen this on occasion in openshift dev CI as well

I've started working on a change related to the comments at https://github.com/openshift/jenkins-client-plugin/issues/209#issuecomment-450855894 and https://github.com/openshift/jenkins-client-plugin/issues/209#issuecomment-451192287

Given the intermittent occurrences for us, I'm a bit worried about validation.  

@szpak or @severi -  if you all are seeing it more readily, and if either of you have cycles / to try pre-release version of my fix when I have it at least functional, please let me know.

thanks

**Comment by gabemontero** [comment]
great - thanks @szpak 

I'll post a zip file of the hpi file here once I've completed my unit / regression testing.

**Comment by gabemontero** [comment]
OK @szpak give this a try:  
[client-plugin.zip](https://github.com/openshift/jenkins-client-plugin/files/2770597/client-plugin.zip)

Unzip the hpi file from that zip and install it

As a secondary concern, if you see issues with this, based on what we see with the job output, I might need to know which version of `oc` you are running with.  There have been some issues from time to time with it's watch processing.

But I suspect this change will help ... this go around I was able to essentially, fully, revert to our pre 1.0.17 file processing.

I'll have PR up shortly to run through our regression suite that will be noted in this issue.

thanks

**Comment by szpak** [comment]
Thanks @gabemontero. I will try to test it on Monday.

**Comment by gabemontero** [comment]
sounds good thanks @szpak

**Comment by gabemontero** [comment]
ok thanks for the info @szpak 

the logging I added did review a couple of things:
1) the output from the watch is not changing
2) the jenkins Proc class thinks the process is still alive

Am I correct in that this particular hung watch is on pods, where you are waiting until the container reaches the ready state?

I *think* I have a couple of more things to look into ... but I also may be providing a new debug jar if those do not pan out if you are still able to try it.

**Comment by maximmold** [comment]
Hey Gabe! Your old buddy Max Moldenhauer here! I think I'm getting the same issue. We are doing an untilEach and at a certain point it stops evaluating the closure. This doesn't always occur, but most of the times it does.

I'm guessing this is due to this change in OcWatch.java?

`if ((newOutput.length > 0 && newOutput.length > outputSize) || firstPass) {
                                firstPass = false;`

I'll evaluate in the morning whether to downgrade or try the patch.

**Comment by bparees** [comment]
@maximmold !

**Comment by szpak** [comment]
> Am I correct in that this particular hung watch is on pods, where you are waiting until the container reaches the ready state?

Yes, it is. And I'm still able to test some other version.

**Comment by maximmold** [comment]
I am getting this with the latest debug plugin.
`07:51:28 
07:51:28 watch closure threw an exception: "groovy.json.JsonException: expecting '}' or ',' but got current char 'a' with an int value of 97
07:51:28 
07:51:28 The current character read is 'a' with an int value of 97
07:51:28 expecting '}' or ',' but got current char 'a' with an int value of 97
07:51:28 line number 14
07:51:28 index number 8199
07:51:28     "apiVersion": "build.openshift.io/v1",
07:51:28 .....^; expecting '}' or ',' but got current char 'a' with an int value of 97
07:51:28 
07:51:28 The current character read is 'a' with an int value of 97
07:51:28 expecting '}' or ',' but got current char 'a' with an int value of 97
07:51:28 line number 14
07:51:28 index number 8199
07:51:28     "apiVersion": "build.openshift.io/v1",
07:51:28 .....^".
07:51:28 `

We are just waiting on a group of parallel builds to complete like so...
`openshift.selector("build", [ 'parent-build' : parentBuildName]).untilEach(3) {
                                echo "Build ${it.name()} ${it.object().status.phase}"
                                if (failed) {
                                    return true
                                }

                                if(it.object().status.phase in ['Failed', 'Cancelled']) {
                                    echo "Build ${it.name()} ${it.object().status.phase} big time!"
                                    failed = true
                                    return true
                                } else if (it.object().status.phase == 'Complete') {
                                    echo "Build ${it.name()} completed now!"
                                    return true
                                }
                            }`

It seems to happen on the completion of the last active build.
I'm going to try reverting to 1.0.22 now.

**Comment by maximmold** [comment]
Reverting to 1.0.22 seems to have resolved my issues.

**Comment by gabemontero** [comment]
OK here is the update for fixing your issue @szpak 
[client-plugin.zip](https://github.com/openshift/jenkins-client-plugin/files/2789082/client-plugin.zip)

@maximmold (good to "talk" with you again Max :-) ) highlighted the offending line ... I got too clever with my optimization (in part because I forgot how we only leverage the `-o=name` option so the output won't change as the pod evovles ... my constant use of `oc get pods -w` messed me up here :-) )

On your `JsonException` in https://github.com/openshift/jenkins-client-plugin/issues/209#issuecomment-456797340 @maximmold ... that is definitely a timing bug, and points to a partial write (or one which had not completed) when we tried to read it  ... a few thoughts

1) not seeing it with 1.0.22 *probably* stems from the fact that version newly employed `Files.readAllBytes`, which we discovered earlier in this issue can block on us, but blocking / reading fully probably worked better in not hitting "partial" json updates (which we have when tailing builds logs 

That blocking call is not viable for us long term

2) I'm curious how consistently you all saw it with your parallel builds ... took a quick stab here and did not see it ... also the fix could have some bearing to the timing ... I'd be curious if it changes the frequency for you all at all, if you all have the bandwidth to try it

3) The same partial read was employed before (up to v1.0.16) and I don't recall issues like this ... but there was substantially more path length with the other jenkins libs used back then .... 

4) I could envision some retry here based on exception type ... once I get the original situation with @szpak and others here sorted out, I can look into that next time I get some cycles for jenkins stuff

thanks

**Comment by szpak** [comment]
It's somehow better @gabemontero . The version `1.0.24-SNAPSHOT (private-3cae71ff-gmontero)` deployed successfully my project 7 of 10 tries (1.0.23 and the previous one was always hanging).

However, I don't see `Checking watch output again in XXXms` any longer, so I'm not sure if it is the same issue. Have you reverted that message intently or I'm somehow testing some old version?

I have spotted some correlation. Here's a fragment of my code with some "debug" statements:
```
while(!checkContainerReadiness(podSelector)) {    //extra loop as I had some issues with multiple pods and untilEach
    echo "Waiting for deployment to be finished (${podSelector.names()})"
    podSelector.untilEach {
        echo "Current status for containers (${it.names()}): ${it.objects()*.status}"
        return checkContainerReadiness(it)
    }
    echo "Status of containers readiness after iteration ${iterationNumber}: ${podSelector.objects()*.status.containerStatuses*.ready}"
}
...
private boolean checkContainerReadiness(def podSelector) {
    println "readiness: ${podSelector.objects()*.status.containerStatuses*.ready}"    //investigating concurrency issues between checks
    boolean areReady = podSelector.objects().every { pod ->
        pod.status.containerStatuses.every {
            it.ready == true
        }
    }
    println "areReady: ${areReady}, readiness: ${podSelector.objects()*.status.containerStatuses*.ready}"
    return areReady
}
```

The timeout seems to occurs (only) in a situation when the container(s) status changes during checkContainerReadiness() method exection:

```
...
12:22:51 Running watch closure body
12:22:54 Current status for containers...
12:22:56 readiness: [[false]]
12:23:00 areReady: false, readiness: [[false]]

12:23:00 Running watch closure body
12:23:08 Current status for containers...
12:23:12 readiness: [[false]]
12:23:16 areReady: false, readiness: [[true]]     <-- readiness changed in the meantime

12:26:01 Cancelling nested steps due to timeout
...
```
Therefore it could be somehow related to my code, but I haven't observed it in 1.0.22.

**Comment by gabemontero** [comment]
Hey @szpak  - glad to hear about the progress.

Continuing with your last comment:

1) The `"Checking watch output again in " + reCheckSleep + "ms having processed " + outputSize + " bytes of watch output so far"` message is still there.  The nature of the latest fix is why you don't necessarily see it repeatedly.  

The fix to date centers on 2  things
a) avoid the blocking file system read
b) make sure to call the closure even if the watch output has not updated (which is entirely possible given we are using `-o=name`)

it was b)  that lead to the unpredictable results.  If the condition in one's watch close returned true  quickly enough, things were good.  However, if not, and the `-o=name` output did not change, we would not call the watch closure again, and it would spin.

2) On your example code, I'm certainly generally try to avoid some of the groovy syntax "sugar" :-) you are employing, but can only remind you what you are most likely very familiar with, that the exit from `podSelector.untilEach` is totally gated by whether your groovy code within the closure returns `true` or `false`.

So with your debug output:
`12:23:16 areReady: false, readiness: [[true]]     <-- readiness changed in the meantime`

since your code is returning `areReady`, which is false, the plugin will keep watching.

In other words, yeah, the difference between 

```
podSelector.objects().every { pod ->
        pod.status.containerStatuses.every {
            it.ready == true
        }
    }
```

and 

```
podSelector.objects()*.status.containerStatuses*.ready
```

is a groovy behavior detail beyond what this plugin can deal with, where any differences are probably timing induced, which might be why you see differences with 1.0.22 and this change.

Just my opinion, but I would avoid fancy groovy and keep it simple.

I'm going to look at some retry for @maximmold 's groovy format exception, and then probably merge the change @szpak has graciously vetted for us.

thanks

**Comment by szpak** [comment]
> The "Checking watch output again in " + reCheckSleep + "ms having processed " + outputSize + " bytes of watch output so far" message is still there. The nature of the latest fix is why you don't necessarily see it repeatedly.

In fact I don't see that message at all. In any of my deployment build. The execution just hangs after "Running watch closure body" (or my debug messages if printed after that). Do you plan to bring it back?

Unfortunately, in the meantime I've got results from some other executions and the watch hangs even if that change does not happen. Therefore, it's most likely not related to my code. It means that in ~30% of cases the watch construction hangs for unknown reason. Do you have any idea why or how to diagnose it?


Regarding the fancy Groovy structured I would like to keep it simple. Unfortunately the JSON structures returned from `oc` are very complicated and sometimes it's problematic to determine if you have any embedded lists/maps or not. The construction with 2 `every` methods is binding, however, it should return the same result - just in different format a list of lists (which is a remainder after I was struggling why I had strange results parsing returned JSON for multipod deployment).
The difference can occur only in a situation a container status (readiness) changed in the meantime. As there is 4 seconds lag between calls it can happen.


Btw, what would be the simplest code to check if all containers in all required replicas are ready?

(as a replacement for:
```
podSelector.objects().every { pod ->
        pod.status.containerStatuses.every {
            it.ready == true
        }
    }
```
)

**Comment by maximmold** [comment]
On 1.0.23, the build was failing to properly complete probably about 75% of the time. I got the JsonException twice in a row so I gave up on that patch. Let me know if there is another fix that I should try out. I'm assuming I'd want one with the retry logic?

**Comment by szpak** [comment]
@gabemontero I will do some more testing on my side tomorrow, as it can be also something in my cluster.

**Comment by gabemontero** [comment]
sounds good @szpak

**Comment by gabemontero** [comment]
went ahead and added the json exception retry @maximmold  ... give it a try at your convenience
[client-plugin-max.zip](https://github.com/openshift/jenkins-client-plugin/files/2793520/client-plugin-max.zip)

**Comment by gabemontero** [comment]
@maximmold I think I may have one more tweak to the patch for you I posted yesterday

**Comment by maximmold** [comment]
@gabemontero Thanks. I should be able to test that out by EOD Monday at the latest. I will try to get to it today if possible though.

**Comment by szpak** [comment]
@gabemontero I tested `client-plugin-dbg1.zip` and it seems that you fixed the original problem. Thanks :+1: 

I have one suspicious related to the occasional timeouts I had. I check if containers are ready. They are not and I return false (inside `untilEach {}`). However, in the meantime the containers status becomes ready, before the control is returned to `untilEach {}`. Would there `untilEach {}` "hang" (my closure is not called anymore) assuming there are no more changes in the pod?

The time slot is very short. In fact I suspect it was happening quite often in my environment as I had:
```
private boolean checkContainerReadiness(def podSelector) {
    boolean areReady = podSelector.objects().every { pod ->
        pod.status.containerStatuses.every {
            it.ready == true
        }
    }
    println "areReady: ${areReady}, readiness: ${podSelector.objects()*.status.containerStatuses*.ready}"
    return areReady
}
```
The second selector call could take a few seconds increasing that time slot. When I changed it to return the value evaluated once more at the end of that method I haven't observed any failure in 6 attempts. 

Therefore, as I already mentioned at the begging, the original problem seems to be fixed. However, I wonder if my conclusion has sense to explain the previous failures in my environment and theoretical chance to have it occurred very rarely after recent changes?

**Comment by gabemontero** [comment]
@szpak:

1) first, good news on confirming the original problem fix

2) to your first question, the status of the pod has no bearing on whether we call the closure; the only reason we would not call the closure are 
  a) the jenkins code we call to read the output file for the underlying `oc get <selector result> -w -o=name` output hangs (jenkins bug)
  b) the calling of the closure has a problem such that it does not reach the contents your provided (jenkins bug)
  c) the `oc get <selector result> -w -o=name` dies unexpectedly (oc bug, openshift bug, flaky environment)

of course, calling the closure and it not returning true is also a possibility for the watch not completing

3) so to your last question, previous environmental issues could have explained some of the unexpected behavior you were previously seeing.

As to moving forward with merging the fix
1) I'm about done with the second iterator for the change for the json exception for @maximmold to try
2) there are a couple of other open issues for the portion of the code I'm going to look at later today ... based on my investigation there, I'll see whether I group those fixes with this change, or cut a new version of  the plugin for this fix first, or lump the changes together and then cut the new version

thanks again

**Comment by maximmold** [comment]
@gabemontero I'm getting the same JsonException. Here is the stack trace.
[jenkinsClientStackTrace.txt](https://github.com/openshift/jenkins-client-plugin/files/2797581/jenkinsClientStackTrace.txt)

**Comment by gabemontero** [comment]
OK, good to know at @maximmold  ... and the stack trace was helpful

But in case it was not clear, the latest patch I attached here does *NOT* have the changes I'm making wrt retry on json exception, etc. and what you are seeing with your parallel builds.

I'm pretty close to unit test / regression test complete on
a) a change that will retry calling 
b) bypasses the jenkins file system stuff (I've confirmed this plugin does not need that for both master and agent based jobs

I hope to have that by EOB today.

Sorry if there was confusion.

**Comment by gabemontero** [comment]
blaaahhh .... `public class JsonException extends RuntimeException`

Gotta change my try/catch from Exception to Throwable will be part of patch I should post soon

**Comment by szpak** [comment]
Nice. I will play with it on Monday.

**Comment by szpak** [comment]
From my side the original problem (100% timed out builds) seems to be fixed and it would be nice to see a new official version in the predictable future.

I will try to find a way to reproduce the occasional issue I was writing about and I will report it in a separate ticket.

**Comment by gabemontero** [comment]
thanks for the update @szpak - current thought is I'll cut a new version this week

@szpak @maximmold  - I had one instance of the json exception in various runs; not sure I am seeing it with enough regularity to be certain, but I have not seen it with this minor tweak 


[client-plugin-max-3.zip](https://github.com/openshift/jenkins-client-plugin/files/2804237/client-plugin-max-3.zip)

**Comment by maximmold** [comment]
Thanks @gabemontero . I will try out the latest one.

**Comment by maximmold** [comment]
@gabemontero The latest patch looks good to me.

**Comment by maximmold** [comment]
@gabemontero I did have one follow up question. The way we are doing our parallel builds requires that you create the pipeline in openshift and then reference them in the coordinating/parent pipeline. We were hoping to avoid defining the pipeline in openshift so people are just using Jenkins, but still get the advantages of using Jenkins slaves running as pods on openshift. Is that a paradigm that makes sense or is supported in any way?

**Comment by gabemontero** [comment]
thanks for the verification on the latest patch @maximmold 

i'll start the process of crafting a new release (which has some atypical hurdles at the moment cause our CI is in flux as a result of 4.0 development)

to your follow up question, yes you can, you just wouldn't use this plugin to start the builds or track their progress.  You can still use this plugin to scope the cluster and project if that makes sense for what you all are thinking, possibly for other non-build interactions with openshift.  Just use the various pipeline directives (either base pipeline, or the dsl provided by the k8s plugin) to specify the agent/slave pods you want to run.

And just in case, moving forward, do make note of some of the restrictions in place when dealing with declarative pipelines and this plugin (https://github.com/openshift/jenkins-client-plugin#compatibility-with-declarative-pipeline)

**Comment by gabemontero** [comment]
v1.0.24 of this plugin has been initiated at the update center
